src/eval.rs

```diff
@@ -172,17 +179,75 @@ impl Default for MiriConfig {
     }
 }
 
-/// Returns a freshly created `InterpCx`, along with an `MPlaceTy` representing
-/// the location where the return value of the `start` function will be
-/// written to.
+/// The state of the main thread. Implementation detail of `on_main_stack_empty`.
+#[derive(Default, Debug)]
+enum MainThreadState {
+    #[default]
+    Running,
+    TlsDtors(tls::TlsDtorsState),
+    Yield {
+        remaining: u32,
+    },
+    Done,
+}
+
+impl MainThreadState {
+    fn on_main_stack_empty<'mir, 'tcx>(
+        &mut self,
+        this: &mut MiriInterpCx<'mir, 'tcx>,
+    ) -> InterpResult<'tcx, Poll<()>> {
+        use MainThreadState::*;
+        match self {
+            Running => {
+                *self = TlsDtors(Default::default());
+            }
+            TlsDtors(state) =>
+                match state.on_stack_empty(this)? {
+                    Poll::Pending => {} // just keep going
+                    Poll::Ready(()) => {
+                        // Give background threads a chance to finish by yielding the main thread a
+                        // couple of times -- but only if we would also preempt threads randomly.
+                        if this.machine.preemption_rate > 0.0 {
+                            *self = Yield { remaining: MAIN_THREAD_YIELDS_AT_SHUTDOWN };
+                        } else {
+                            *self = Done;
+                        }
```

Comment:
This is still the yield-based impl, not the step-based idea. I wouldn't mind saving the step-based impl for a new PR, but I just want to make sure: Did you use yielding here because the step-based idea is too hard now? Or just because the yield strategy is what was in my branch?

Reply:
I think yielding is sufficient actually -- note how I only yield if preemption is enabled. So (a) there's anyway a chance that we might have scheduled to those other threads without this extra yield, and (b) there's also a chance we will schedule back to the main thread often enough to eventually terminate the program. Once we are in the `Yield` state, Miri is *guaranteed* to terminate, though with very small preemption rates it can take very long.

Reply:
And there's no way we could get these unhelpful leak reports with preemption disabled?

Reply:
I don't think so, no. The scoped thread logic would have to deliberately yield somewhere in the code that runs after the user thread is done.